src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2085,7 +2122,13 @@ public new SemanticModel GetSemanticModel(SyntaxTree syntaxTree, bool ignoreAcce
                 throw new ArgumentException(CSharpResources.SyntaxTreeNotFound, nameof(syntaxTree));
             }
 
-            return new SyntaxTreeSemanticModel(this, (SyntaxTree)syntaxTree, ignoreAccessibility);
+            if (SemanticModelProvider != null && useSemanticModelProviderIfNonNull)
+            {
+                Debug.Assert(!ignoreAccessibility);
```

Comment:
This is another reason to prefer two methods: `GetSemanticModelCore` and `CreateSemanticModel`. The input to the single method is only conditionally correct. Removes an entire class of errors if we have two separate methods here.